optimizely/helpers/condition.py

```diff
@@ -12,115 +12,180 @@
 # limitations under the License.
 
 import json
+import numbers
 
+from six import string_types
 
-class ConditionalOperatorTypes(object):
+from .validator import is_finite_number, are_values_same_type
+
+
+class ConditionOperatorTypes(object):
   AND = 'and'
   OR = 'or'
   NOT = 'not'
 
 
-DEFAULT_OPERATOR_TYPES = [
-  ConditionalOperatorTypes.AND,
-  ConditionalOperatorTypes.OR,
-  ConditionalOperatorTypes.NOT
-]
+class ConditionMatchTypes(object):
+  EXACT = 'exact'
+  EXISTS = 'exists'
+  GREATER_THAN = 'gt'
+  LESS_THAN = 'lt'
+  SUBSTRING = 'substring'
+
 
+class CustomAttributeConditionEvaluator(object):
+  """ Class encapsulating methods to be used in audience leaf condition evaluation. """
 
-class ConditionEvaluator(object):
-  """ Class encapsulating methods to be used in audience condition evaluation. """
+  CUSTOM_ATTRIBUTE_CONDITION_TYPE = 'custom_attribute'
 
   def __init__(self, condition_data, attributes):
     self.condition_data = condition_data
-    self.attributes = attributes
+    self.attributes = attributes or {}
 
-  def evaluator(self, condition):
-    """ Method to compare single audience condition against provided user data i.e. attributes.
+  def is_value_valid_for_exact_conditions(self, value):
+    """ Method to validate if the value is valid for exact match type evaluation
 
     Args:
-      condition: Integer representing the index of condition_data that needs to be used for comparison.
+      value: Value to validate
 
     Returns:
-      Boolean indicating the result of comparing the condition value against the user attributes.
+      Boolean: True if value is a string type, or a boolean, or is finite. Otherwise False
     """
+    if isinstance(value, string_types) or isinstance(value, bool) or is_finite_number(value):
+      return True
 
-    return self.attributes.get(self.condition_data[condition][0]) == self.condition_data[condition][1]
+    return False
 
-  def and_evaluator(self, conditions):
-    """ Evaluates a list of conditions as if the evaluator had been applied
-    to each entry and the results AND-ed together
+  def exact_evaluator(self, index):
+    """ Evaluate the given exact match condition for the user attributes
 
     Args:
-      conditions: List of conditions ex: [operand_1, operand_2]
+      index: Index of the condition to be evaluated
```

Comment:
index is a built in Python method I think. Please rename to something else.

Reply:
Objects have a special method named `__index__`, but no global `index` variable, I don't think. See the global functions reference in [2](https://docs.python.org/2/library/functions.html) and [3](https://docs.python.org/3/library/functions.html).